### ISSUE TITLE
Issues with "as" vs "is" props in V3

### ISSUE THREAD

**Issue by ntaylor89** [rebassjs/rebass#571]
I noticed that the docs recommend using `as` to change the underlying tag emitted. However, if I use `as` instead of is, then all of the `styled-system` helpers are removed, and the values are passed directly to the DOM.

i.e. this works

```jsx
const Code = props => 
  <Text
    is='code'
    fontFamily='mono'
  />
```

However, if we swap `is` with `as`, `fontFamily` now gets passed as `font-family` as an attribute on the tag directly. I get

```jsx
  <code font-family="mono" class="random-gen-class-name"></code>
```

**Comment by jxnblk** [comment]
What version of styled-components are you using?

**Comment by ntaylor89** [comment]
4.1.3.

Some background: I'm trying to make a component lib in it's own package. I'm using `x0` to dev, which is where I'm seeing this behavior (haven't tried anywhere else yet).

I've patched `x0` to support the latest babel (so I could use the `styled-components` babel plugin).

Main dependencies are:

```
"rebass": "^3.0.1",
"styled-components": "^4.1.3"
```

I know that `x0` has `rebass` and probably `styled-components` as dependencies. I don't know if maybe those are also conflicting?

**Comment by mrfelton** [comment]
I'm getting the same issue:

```
"babel-plugin-styled-components": "1.10.0",
"styled-components": "4.1.3",
"rebass": "3.0.1",
```

**Comment by mrfelton** [comment]
In my case, swapping `as` for `is` does not seem to have an effect. With `as` or `is` in place, none of the system helper props work.

**Comment by mrfelton** [comment]
Also worth noting, this only seems to be an issue for me if there is a `css` prop.

eg:


This works as expected
```js
<Box as="aside" p={5} m={5} />
```

So does this:
```js
<Box p={5} m={5} css="overflow: hidden;" />
```

But when both `css` and `as` are applied, the `p` and `m` are not applied (neither are any other helper props from `styled-system`.
```js
<Box as="aside" p={5} m={5} css="overflow: hidden;" />
```

**Comment by gi-alec** [comment]
@mrfelton I've been experiencing similar unexpected outputs when using the "as" prop in addition to extending a rebass component (using `styled(Component)`), as well as the case you mentioned. I've grown to fear the `as` prop, and regrettably as a result my website has turned out to be quite the div soup.

**Comment by jxnblk** [comment]
> I know that x0 has rebass and probably styled-components as dependencies. I don't know if maybe those are also conflicting?

Those cross dependencies *are* (almost) definitely affecting the output. I'd recommend trying either with a basic webpack setup or using something like Gatsby or Next.js. It might also be worth trying the Emotion version out if possible as that library tends to have fewer issues with npm dependencies

As far as the `font-family` attribute goes, I think that is one of the props that styled-components whitelists as valid HTML – see this issue: https://github.com/styled-components/styled-components/issues/439

**Comment by mrfelton** [comment]
@jxnblk do you have any ideas what's causing this or how to resolve? Removing all `as` props from our codebase isn't a great option.

**Comment by jxnblk** [comment]
The best way to handle this currently is to use Emotion and the `shouldForwardProp` API. There's a utility here for styled-system props https://github.com/styled-system/extras/tree/master/packages/should-forward-prop

This is a long-standing issue in styled-components that still hasn't been resolved, see https://github.com/styled-components/styled-components/issues/439

**Comment by mrfelton** [comment]
Thanks, but switching from `styled-components` to `emotion` is also not a good option!

**Comment by jxnblk** [comment]
I totally get that, but the issue is really a styled-components issue and trying to solve it in user land has historically caused a lot of other issues

**Comment by karlhorky** [comment]
Maybe this should be either documented in the `rebass` docs or the `as` change reverted until `styled-components` fixes their issue?

> trying to solve it in user land has historically caused a lot of other issues

Or do the other issues with providing `as` in the API again outweigh this? What are the issues @jxnblk?

**Comment by mrfelton** [comment]
@jxnblk Does the fix in https://github.com/styled-components/styled-components/pull/2573 provide what you need to resolve this issue? Or, would that fix by itself resolve this?

**Comment by aroc** [comment]
@mrfelton's solution worked for me (pinning the versions to the ones he suggests).

**Comment by mrfelton** [comment]
The latest versions of `styled-components`, `babel-plugin-styled-components` and `rebass` work cleanly together now, so this appears to no longer be an issue.

**Comment by karlhorky** [comment]
Oh cool, which versions are those? Do you know where the fix came from?

I guess if we can identify that the original problem is indeed fixed with some commit on one or more of those projects, then this issue can be closed.

**Comment by mrfelton** [comment]
>  which versions are those

latest:
```
"babel-plugin-styled-components": "1.10.0",
"styled-components": "4.3.1",
"rebass": "3.1.1",
```
> Do you know where the fix came from?

See my previous comment in this thread - https://github.com/rebassjs/rebass/issues/571#issuecomment-498966290

**Comment by mrfelton** [comment]
Note, you will need to convert your css props to use the new styled-components native css prop format.

**Comment by karlhorky** [comment]
Thanks for the info! This will be very helpful for anyone else running into this issue.

> the new styled-components native css prop format

Not sure if I'm sure what you mean here. Do you have a link for this?

**Comment by mrfelton** [comment]
https://www.styled-components.com/docs/api#css-prop

This is different from the old rebass css prop.

Old style (expects css object):
```js
css={{
  'background-color': 'red',
  'border': '1px solid green';
}}
```

New style (allows you to write actual css):
```js
css={`
  background-color: red;
  border: 1px solid green;
`}
```

**Comment by jxnblk** [comment]
Closing this due, since I *think* the original issue has been covered

**Comment by karlhorky** [comment]
> The best way to handle this currently is to use Emotion and the shouldForwardProp API. There's a utility here for styled-system props styled-system/extras:packages/should-forward-prop@master 
> 
> This is a long-standing issue in styled-components that still hasn't been resolved, see styled-components/styled-components#439

This has now been resolved! `styled-components@5.1.0` has Emotion-style `shouldForwardProp` support:

https://github.com/styled-components/styled-components/releases/tag/v5.1.0

Original PR:

https://github.com/styled-components/styled-components/pull/3006

---

I've opened #953 to investigate using this new API in Rebass.
